bin/ch-completion.bash

```diff
@@ -0,0 +1,403 @@
+#     https://www.gnu.org/software/bash/manual/html_node/Shell-Parameter-Expansion.html
+#
+#   * Bash completion builtins (compgen, comopt, etc.):
+#     https://www.gnu.org/software/bash/manual/html_node/Programmable-Completion-Builtins.html
+#
+#   * Bash completion variables (e.g. COMPREPLY):
+#     https://devmanual.gentoo.org/tasks-reference/completion/index.html
+#
+#   * Call-by-reference for bash function args:
+#     https://unix.stackexchange.com/a/224564
+#
+#
+## SYNTAX GLOSSARY ##
+#
+# This script uses syntax that may be confusing for bash newbies and those who
+# are rusty.
+#
+# Source:
+# https://www.gnu.org/software/bash/manual/html_node/Shell-Parameter-Expansion.html
+#
+# ${array[i]}
+#   Gives the ith element of “array”. Note that bash arrays are indexed
+#   at zero, as all things should be.
+#
+# ${array[@]}
+#   Expands “array” to its member elements as a sequence of words, one word per
+#   element.
+#
+# ${#parameter}
+#   Gives the length of “parameter.” If “parameter” is a string, this expansion
+#   gives you the character length of the string. If “paramter” is an array
+#   subscripted by “@” or “*” (e.g. “foo[@]”), then the expansion gives you the
+#   number of elements in the array.
+#
+# ${parameter:offset:length}
+#   A.k.a. substring expansion. If “parameter” is a string, expand up to
+#   “length” characters, starting with the character at position “offset.” If
+#   “offset” is unspecified, start at the first character. If “parameter” is an
+#   array subscripted by “@” or “*,” (e.g. “foo[@]”) expand up to “length”
+#   elements, starting at the element at position “offset” (e.g.
+#   “${foo[offset]}”).
+#
+#   Example 1 (string):
+#   $ foo="abcdef"
+#   $ echo ${foo::3}
+#   abc
+#   $ echo ${foo:1:3}
+#   bcd
+#
+#   Example 2 (array):
+#   $ foo=("a" "b" "c" "d" "e" "f")
+#   $ echo ${foo[@]::3}
+#   a b c
+#   $ echo ${foo[@]:1:3}
+#   b c d
+#
+
+# Possible extensions once this is merged:
+#   * Add completion support for non-bash shells (e.g. zsh and tchs).
+#     see https://github.com/git/git/tree/master/contrib/completion
+#   * Add support for mid-line completion (possibly using COMP_POINT,
+#     see https://devmanual.gentoo.org/tasks-reference/completion/index.html).
+#
+
+# Minimum supported bash version. According to
+# http://mywiki.wooledge.org/BashFAQ/061 and my own testing, negative array
+# indexing was introduced in this version. To simplify confusing syntax, we this
+# script is unsupported for bash < 4.2.0.
+bash_vmin="4.2.0"
+
+# Check bash version
+bash_v=$(bash --version | head -1 | grep -Eo "[0-9\.]{2,}[0-9]")
+if [[ $(printf "%s\n%s\n" "$bash_vmin" "$bash_v" | sort -V) < "$bash_vmin" ]]; then
+    echo "ch-completion.bash: unsupported bash version ($bash_v < $bash_vmin)"
+    return
+fi
+
+# Debugging log
+if [[ -f "/tmp/ch-completion.log" && -n "$CH_COMPLETION_DEBUG" ]]; then
+    rm /tmp/ch-completion.log
+    printf "completion log\n\n" >> /tmp/ch-completion.log
```

Comment:
You probably want to always append so that people can compare results from different sourcings.